### PR TITLE
Bug simplify algegra

### PR DESCRIPTION
**Pull request by scxiao**

### CODE REVIEW

File: src/simplify_algebra.cpp [hunk 1]
```diff
@@ -439,6 +444,39 @@ struct find_splits
         return groups;
     }
 
+    std::vector<int64_t> get_reduce_axes(instruction_ref ins) const
+    {
+        assert(contains(ins->name(), "reduce"));
+        std::vector<int64_t> axes;
```
Comment (pfultz2) [inline src/simplify_algebra.cpp:450]: This should just read the axes with `ins->get_operator().to_value()["axes"].to_vector<int64_t>()`.

File: src/simplify_algebra.cpp [hunk 2]
```diff
@@ -453,6 +491,23 @@ struct find_splits
             if(op.name() == "slice")
                 continue;
 
+            if(contains(op.name(), "reduce"))
```
Comment (pfultz2) [inline src/simplify_algebra.cpp:494]: If we are going to support reduce then the matcher should be updated to match reduce.

File: src/simplify_algebra.cpp [hunk 3]
```diff
@@ -453,6 +491,23 @@ struct find_splits
...
+                auto reduce_axes = get_reduce_axes(start);
+                // axes of slice and reduce op cannot have overlap
+                if(std::any_of(slc_axes.begin(), slc_axes.end(), [&](auto axis) {
+                       return (std::find(reduce_axes.begin(), reduce_axes.end(), axis) !=
+                               reduce_axes.end());
+                   }))
+                {
+                    continue;
+                }
+            }
```
Comment (pfultz2) [inline src/simplify_algebra.cpp:509]: We need and else here to check that the operator is pointwise. If its not reduce or pointwise then we should `continue`.

File: src/simplify_algebra.cpp [hunk 4]
```diff
@@ -406,7 +406,7 @@ struct find_splits
     auto matcher() const
     {
         return match::any(match::any_of[match::outputs()](
-            match::name("slice")(match::any_of[match::outputs()](pointwise()))));
+            match::name("slice")(match::any_of[match::outputs()](pointwise(), match::name_contains("reduce")))));
```
Comment (pfultz2) [inline src/simplify_algebra.cpp:409]: Just like we have a `pointwise` matcher(defined [here](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/2ab73fd32145fa645da05c0e6e2fee77cd1889a0/src/simplify_algebra.cpp#L43)), could we add a `reduction` matcher to check for reduce operators?

**Review by pfultz2 — APPROVED** [review]

**Review by kahmed10 — APPROVED** [review]

**Comment by scxiao** [comment]
Looks like all checks have passed, but it does not display here correctly. So this PR can be merged now.

**Pull request merged**